docs: Add zustand store action naming conventions dev guide (fixes #289). (#332)
y-scope/yscope-log-viewer: docs/src/dev-guide/coding-guidelines.md

the state variable with an additional `Ref` suffix. E.g., `logEventNumRef` is the reference variable
that corresponds to the `logEventNum` state variable.

# Zustand

## File Naming

When creating Zustand stores, we follow these naming conventions:

* Store files: `{name}Store.ts` (camelCase).
* Large stores: create `{name}Store/` folder with:
  * `index.ts` - main store file that combines slices.
  * `create{Name}{Feature}Slice.ts` - individual feature slices (e.g., `createQueryConfigSlice.ts`).

## Store structure

### Type definitions

Split store types into three interfaces:
* `{Name}Values` - state variables
* `{Name}Actions` - action functions  
* `{Name}State` or `{Name}Slice` - combined type

```ts
interface LogExportValues {
    exportProgress: Nullable<number>;
}

interface LogExportActions {
    setExportProgress: (newProgress: Nullable<number>) => void;
}

type LogExportState = LogExportValues & LogExportActions;

> 

Author: ```suggestion
```

```

```

### Default values

We define default values for state variables during the creation.
The default values should be in CAPITAL_SNAKE_CASE. Here's the continuation of the previous example:

```ts
const LOG_EXPORT_STORE_DEFAULT: LogExportValues = {
    exportProgress: null,
};
```

Notice that we define `LOG_EXPORT_STORE_DEFAULT` using `LogExportValues` interface, which can help
type check the default values.

### Actions Naming

When implementing Zustand store actions, we follow these naming conventions:

* Use `setXxx` for actions that simply change the value of a state.
* Use `updateXxx` for actions that do more than simply changing a value (e.g., perform additional logic, make API calls,
update multiple states).

## Feature-based slicing

When a Zustand store file becomes too large, we should slice it based on features. Avoid grouping by type (e.g., all
values / actions in one object) — it's a common anti-pattern.

## Store Usage

There are three ways to access Zustand store values and actions:
* `get() & set()`
* `const yyy = useXxxStore((state) => state.yyy);`
* `const {yyy, zzz} = useXxxStore.getState();`

### Inside zustand creation

> When accessing store values inside the store creation function, we use`set()` and `get()`. Here's an example:

Author: ```suggestion
When accessing store values inside the store creation function, we use `set()` and `get()`. Here's an example:
```

```ts
const useLogExportStore = create<LogExportState>((get, set) => ({
    exportLogs: () => {
      const logExportManager = new LogExportManager();
      set({logExportManager});
      
      const {exportProgress} = get();
      // If name doesn't match, use colon to separate the variable name from the value.
      set({exportProgress: EXPORT_LOGS_PROGRESS_VALUE_MIN});
    },
}));
```

> ### Inside React components

Reviewer: To me, it makes more sense to list the "State values" sections first, which includes the only scenario where reactive access is used

> #### State variables
> `const vvv = useXxxStore((state) => state.vvv);`
> 
> This will assure whenever `vvv` changes, the component will re-render. For variables that are not subscribing to changes,
> use `const {vvv} = useXxxStore.getState()`; instead. This avoids unnecessary dependencies in react hooks.
> 
> #### Actions
> `const {aaa} = useXxxStore.getState();`
> 
> Since actions don't change after initialization, we can access them in a non-reactive way.

Reviewer: ```suggestion
#### Actions

Actions usually do not change after initialization, so always access them non-reactively:
```ts
const handleExportButtonClick = useCallback(() => {
    const {exportLogs} = useLogExportStore.getState();
    exportLogs();
}, []);
```

#### State values

Choose access pattern based on usage:

*Reactive access* - when the value is used in JSX or hook dependency arrays:
```ts
const exportProgress = useLogExportStore((state) => state.exportProgress));

// The progress should be printed when `exportProgress` updates.
useEffect(() => {
    console.log(exportProgress);
}, [exportProgress]);
```

*Non-reactive access* - when the value should not trigger re-renders or hook re-runs:
```ts
// The progress should be printed only once when the component mounts.
useEffect(() => {
    const {exportProgress} = useLogExportStore.getState();
    console.log(exportProgress);
}, []);
```
```

### Outside React components

> When accessing Zustand store values or actions outside React components, we use
> `const {yyy, zzz} = useXxxStore.getState();`.

Reviewer: ```suggestion
Always use non-reactive access since reactive subscriptions do not work outside components.
```

}

type LogExportState = LogExportValues & LogExportActions;
```

### Default values

* Create an object for initial state values.
* Type with `{Name}Values` interface for validation.

```ts
const LOG_EXPORT_STORE_DEFAULT: LogExportValues = {
    exportProgress: null,
};
```

### Action naming

* `set{Property}` - simple state updates.
* `update{Property}` - complex logic, API calls, or multiple state updates.

```ts
const useUserStore = create<UserState>((set, get) => ({
    setName: (name) => {
        set({name});
    },
    updateProfile: async (data) => {
        set({isLoading: true});
        const result = await api.updateProfile(data);
        set({
            profile: result,
            isLoading: false
        });
    },
}));
```

## Feature-based slicing

When a Zustand store file becomes too large, we should slice it based on features. Avoid slicing by type (e.g., all
values / actions in one object) - it's a common anti-pattern.

## Store access patterns

### Inside store creation

Use `get()` and `set()` to access the store's own states:

```ts
const useLogExportStore = create<LogExportState>((get, set) => ({
    exportLogs: () => {

>       const logExportManager = new LogExportManager();
>       set({logExportManager});
>       

Reviewer: ```suggestion
    // ...
```

>       const {exportProgress} = get();

Reviewer: it doesn't show how the progress is used once retrieved, right? can we come up with a better example?

Author: Do you mean we should replace this whole example and come up with some examples using other stores?

I think this is fine because it demonstrate how the exportProgress is get; how it would be used is kinda out of our purpose. Plus, there's no difference using this extracted variable compared to other variables.

Reviewer: that we retrieve `exportProgress` and do nothing with it is a bit odd - the sample looks a bit incomplete. It might also confuse people who are unfamiliar with the destructuring syntax.

Do you think the example proposed at https://github.com/y-scope/yscope-log-viewer/pull/332#discussion_r2175758344 is a better one?

the state variable with an additional `Ref` suffix. E.g., `logEventNumRef` is the reference variable
that corresponds to the `logEventNum` state variable.

# Zustand

## File Naming

When creating Zustand stores, we follow these naming conventions:

* Store files: `{name}Store.ts` (camelCase).
* Large stores: create `{name}Store/` folder with:
  * `index.ts` - main store file that combines slices.
  * `create{Name}{Feature}Slice.ts` - individual feature slices (e.g., `createQueryConfigSlice.ts`).

## Store structure

### Type definitions

Split store types into three interfaces:

* `{Name}Values` - state variables
* `{Name}Actions` - action functions
* `{Name}State` - union of values and actions

> When store gets too large and exceeds eslint `max-lines-per-function limit`, we slice `{Name}State`
> to `{Name}{Feature}Slice`. Then we unionize these slices to `{Name}State`.

Author: ```suggestion
When the store becomes too large and exceeds the ESLint `max-lines-per-function` limit, split
`{Name}State` into feature-specific slices (`{Name}{Feature}Slice`) and then combine those slices
back into `{Name}State`.
```

the state variable with an additional `Ref` suffix. E.g., `logEventNumRef` is the reference variable
that corresponds to the `logEventNum` state variable.

# Zustand

## File Naming

When creating Zustand stores, we follow these naming conventions:

* Store files: `{name}Store.ts` (camelCase).
* Large stores: create `{name}Store/` folder with:
  * `index.ts` - main store file that combines slices.
  * `create{Name}{Feature}Slice.ts` - individual feature slices (e.g., `createQueryConfigSlice.ts`).

## Store structure

### Type definitions

Split store types into three interfaces:

* `{Name}Values` - state variables
* `{Name}Actions` - action functions
* `{Name}State` - union of values and actions

> When store gets too large, we slice `{Name}State` to `{Name}{Feature}Slice`.

Reviewer: shall we define "too large"?

the state variable with an additional `Ref` suffix. E.g., `logEventNumRef` is the reference variable
that corresponds to the `logEventNum` state variable.

# Zustand

> 

Reviewer: For now, let's modify only the documents in this PR. I agree adding an introduction will help the text flow
